mltensor-explainer.md

```diff
@@ -0,0 +1,423 @@
+
+### Allow mapping an `MLTensor` to script
+
+Why doesn't `MLTensor` have a [`mapAsync()`](https://www.w3.org/TR/webgpu/#dom-gpubuffer-mapasync) method, as a `GPUBuffer` does?
+
+`MLTensor`s are used as inputs and outputs to machine learning models. In practice, we expect the size of the model inputs and outputs to be dwarfed by the size of the model weights, which are uploaded via the `MLGraphBuilder.constant()` method. We may re-evaluate this stance in the future if we discover that reading and writing data to an `MLTensor` is a bottleneck, though even in that case we may prefer to explore a solution which bypasses `ArrayBuffer`s altogether, as discussed [above](#support-passing-an-mltensor-as-a-constant).
+
+### Hash input buffer contents
+
+One approach to solve the [buffer reuse](#buffer-reuse) case is for the WebNN implementation to silently avoid making redundant data copies if the same buffer contents are repeatedly passed as inputs. This may be achieved by hashing the contents of each input. This approach has downsides, such as [managing the extra copies in the hash map](#resource-management) and that hashing the buffer contents may be expensive as it requires reading the entire input. This approach also does not address the other use cases.
+
+## References & Acknowledgements
+
+Many thanks for valuable feedback and advice from:
+
+- Bryan Bernhart
+- Joshua Bell
+- Mike Wyrzykowski
+- Ningxin Hu
+- Phillis Tang
+- Rafael Cintron
+- Reilly Grant
+- Zoltan Kis
+
+---
+
+## Appendix
+
+### Tentative IDL
+
+```javascript
+typedef [EnforceRange] unsigned long MLTensorUsageFlags;
+
+namespace MLTensorUsage {
+    const MLFlagsConstant READ_FROM      = 0x0001;
+    const MLFlagsConstant WRITE_TO       = 0x0002;
+    const MLFlagsConstant WEBGPU_INTEROP = 0x0004;
+};
+
+dictionary MLTensorDescriptor : MLOperandDescriptor {
+  required MLTensorUsageFlags usage;
+};
+
+typedef record<DOMString, MLTensor> MLNamedTensors;
+
+interface MLTensor {
+  readonly attribute MLOperandDataType dataType;
+  readonly attribute FrozenArray<unsigned long> shape;
+  readonly attribute unsigned long MLTensorUsageFlags usage;
+
+  void destroy();
+};
+
+partial interface MLContext {
+  Promise<MLTensor> createTensor(MLTensorDescriptor descriptor);
+
+  void writeBuffer(MLTensor dstTensor, [AllowShared] ArrayBuffer srcData);
+  void writeBuffer(MLTensor dstTensor, [AllowShared] ArrayBufferView srcData);
+  
+  Promise<ArrayBuffer> readBuffer(MLTensor srcTensor);
+  Promise<void> readBuffer(MLTensor srcTensor, [AllowShared] ArrayBuffer dstData);
+  Promise<void> readBuffer(MLTensor srcTensor, [AllowShared] ArrayBufferView dstData);
+  
+  void dispatch(MLGraph graph, MLNamedTensors inputs, MLNamedTensors outputs);
+};
+
+// For WebGPU Interop
+
+interface GPUExternalBuffer {};
+GPUExternalBuffer includes GPUObjectBase;
```

Comment:
Depending on how opaque the internal data is for WGSL shaders (see other comment) this wouldn't be an external buffer, but more like a `GPUImportedMLTensor` maybe? Buffer makes it sound like the layout would be transparent and the buffer could be seen as an `array<u32>` directly.

Reply:
See my response on the other comment. If this object is indeed addressable like a generic `GPUBuffer` then is `GPUExternalBuffer` reasonable or would you prefer the name indicate that it came from WebNN (by using "ML" or "tensor")?

Reply:
Updated to use `array<T>` so we can just import as a `GPUBuffer`. Longer-term, it would be nice to have some sort of `GPUImportedTensor` type to abstract away the layout of the buffer, but we'll start with this